### PR TITLE
Add invoke cheatcode

### LINKED ISSUE #1348: Add `invoke` cheatcode

### LINKED ISSUE #1360: Add more tests for `invoke` cheatcode
After `call` is merged and it's possible to query the state

### PR DESCRIPTION
**Pull request by Arcticae**
Closes #1348 
closes #1360

### CODE REVIEW

**Comment by Arcticae** [comment]
Related #1360

File: protostar/starknet/cheaters/contracts.py
Comment (karol-bisztyga) [inline protostar/starknet/cheaters/contracts.py:290]: I think this should be moved to a separate function, I did this in #1358 so I'd wait for that PR to be merged and hen rebase, what do you think?

File: tests/integration/pure_cairo_vm/cheatcodes/invoke/invoke_test.cairo
Comment (Arcticae) [reply tests/integration/pure_cairo_vm/cheatcodes/invoke/invoke_test.cairo:10]: yes that's true, want to do it in #1360

File: protostar/starknet/cheaters/contracts.py
Comment (Arcticae) [reply protostar/starknet/cheaters/contracts.py:290]: sure, we can wait for that one

File: protostar/testing/cairo_cheatcodes/invoke_cairo_cheatcode.py
Comment (karol-bisztyga) [inline protostar/testing/cairo_cheatcodes/invoke_cairo_cheatcode.py:28]: Nit: For other cheat codes we add a separate private method where we catch `ContractsCheaterException` and rethrow it as `CheatcodeException`. I think we should be consistent with this as long as you're not 100% sure that there's no way here for `ContractsCheaterException` to occur.

File: tests/integration/pure_cairo_vm/cheatcodes/invoke/invoke_test.cairo
Comment (MaksymilianDemitraszek) [inline tests/integration/pure_cairo_vm/cheatcodes/invoke/invoke_test.cairo:3]: Please test it through proxy contract

File: tests/integration/pure_cairo_vm/cheatcodes/invoke/invoke_test.cairo
Comment (karol-bisztyga) [inline tests/integration/pure_cairo_vm/cheatcodes/invoke/invoke_test.cairo:14]: Nit: this should probably be right after
```
%{ ids.deployed_contract_address = deploy_contract("./src/basic.cairo").contract_address %}
```

File: tests/integration/pure_cairo_vm/cheatcodes/invoke/invoke_test.cairo
Comment (karol-bisztyga) [inline tests/integration/pure_cairo_vm/cheatcodes/invoke/invoke_test.cairo:23]:
```suggestion
    local new_balance;
    %{ invoke(ids.deployed_contract_address, "increase_balance", [123]) %}
    %{
        ids.new_balance = call(ids.deployed_contract_address, "get_balance")[0]
    %}
    assert new_balance = 223;

```

**Review by karol-bisztyga — APPROVED** [review]

File: tests/integration/pure_cairo_vm/cheatcodes/invoke/invoke_test.cairo
Comment (MaksymilianDemitraszek) [reply tests/integration/pure_cairo_vm/cheatcodes/invoke/invoke_test.cairo:3]: We should do the same for the call cheatcode

@karol-bisztyga can you please add such tests?

**Review by MaksymilianDemitraszek — COMMENTED** [review]
Looks good but please test through proxy, historically there were some bugs including that

**Review by MaksymilianDemitraszek — APPROVED** [review]
Please add proxy tests before merging

File: tests/integration/pure_cairo_vm/cheatcodes/invoke/invoke_test.cairo
Comment (Arcticae) [reply tests/integration/pure_cairo_vm/cheatcodes/invoke/invoke_test.cairo:3]: #1361

**Pull request merged**
